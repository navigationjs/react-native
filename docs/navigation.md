---
id: navigation
title: Navigation
---

## Example

Example

### `name`

Name
