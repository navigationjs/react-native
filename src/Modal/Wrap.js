import React, { Component } from 'react';
import { Animated, Dimensions } from 'react-native';

const { height } = Dimensions.get('window');

export default class Wrap extends Component {
  state = { loading: this.props.scene.active.loading };

  componentDidMount() {
    const {
      scene: { active },
    } = this.props;
    active.value.addListener(() => this.setState({ loading: active.loading }));
  }

  render() {
    const { scene, children, overlay, style, ...props } = this.props;
    const { loading } = this.state;

    return (
      <>
        {overlay && (
          <Animated.View
            style={{
              position: 'absolute',
              height: '100%',
              width: '100%',
              overflow: 'hidden',
              transform: [
                {
                  translateY: scene.active.value.interpolate({
                    inputRange: [0, 0, 1],
                    outputRange: [height, 0, 0],
                  }),
                },
              ],
              backgroundColor: scene.active.value.interpolate({
                inputRange: [0, 1],
                outputRange: ['rgba(0,0,0,0)', 'rgba(0,0,0,0.4)'],
              }),
            }}
          />
        )}
        <Animated.View
          {...props}
          style={[
            {
              position: 'absolute',
              width: '100%',
              height: '100%',
              backgroundColor: 'white',
              overflow: 'hidden',
            },
            {
              transform: [
                {
                  translateY: scene.active.value.interpolate({
                    inputRange: [0, 1],
                    outputRange: [height, 0],
                  }),
                },
              ],
            },
            style,
          ]}
        >
          {typeof children === 'function' ? children({ loading }) : children}
        </Animated.View>
      </>
    );
  }
}
