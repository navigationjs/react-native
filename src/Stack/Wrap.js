import React, { Component } from 'react';
import { Animated, Dimensions } from 'react-native';

const { width } = Dimensions.get('window');

export default class Wrap extends Component {
  state = { loading: this.props.scene.active.loading };

  componentDidMount() {
    const {
      scene: { active },
    } = this.props;
    active.value.addListener(() => this.setState({ loading: active.loading }));
  }

  render() {
    const { scene, children, style, ...props } = this.props;
    const { loading } = this.state;

    return (
      <Animated.View
        {...props}
        style={[
          {
            position: 'absolute',
            width: '100%',
            height: '100%',
            backgroundColor: 'white',
            overflow: 'hidden',
          },
          {
            transform: [
              {
                translateX: scene.active.value.interpolate({
                  inputRange: [0, 1],
                  outputRange: [width, 0],
                }),
              },
              {
                translateX: scene.depth.value.interpolate({
                  inputRange: [0, 1],
                  outputRange: [0, -width / 2],
                }),
              },
            ],
          },
          style,
        ]}
      >
        {typeof children === 'function' ? children({ loading }) : children}
      </Animated.View>
    );
  }
}
