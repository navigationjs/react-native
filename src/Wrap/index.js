import React, { Component } from 'react';
import { View, BackHandler } from 'react-native';
import Back from '../Back';

export const link = { wrap: null };

export default class Wrap extends Component {
  state = { disabled: false };

  componentDidMount() {
    link.wrap = this;

    this.backHandler = BackHandler.addEventListener(
      'hardwareBackPress',
      this.handleBackPress
    );
  }

  componentWillUnmount() {
    link.wrap = null;
    this.backHandler.remove();
  }

  handleBackPress = () => {
    if (!this.state.disabled) Back.dispatch();
    return true;
  };

  enable = () => this.setState({ disabled: false });
  disable = () => this.setState({ disabled: true });

  render() {
    const { children, ...props } = this.props;
    const { disabled } = this.state;

    return (
      <View
        {...props}
        style={{ flex: 1 }}
        pointerEvents={disabled ? 'none' : 'auto'}
      >
        {children}
      </View>
    );
  }
}
