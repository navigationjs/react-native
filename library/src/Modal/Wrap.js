import React, { Component } from 'react';
import { Animated, Dimensions } from 'react-native';
import { toId } from '../helpers';
import navigation from '../Navigation';

const { height } = Dimensions.get('window');

export default class Wrap extends Component {
  constructor(props) {
    super(props);

    const { navigator: navigatorName, scene: sceneName } = props;

    const navigator = navigation.navigators[navigatorName];
    const scene = navigator.scenes[sceneName];

    this.state = { loading: scene.active.loading };
  }

  componentDidMount() {
    const { navigator: navigatorName, scene: sceneName } = this.props;
    const navigator = navigation.navigators[navigatorName];
    const scene = navigator.scenes[sceneName];
    scene.active.value.addListener(() =>
      this.setState({ loading: scene.active.loading })
    );
  }

  render() {
    const {
      navigator: navigatorName,
      scene: sceneName,
      children,
      overlay,
      style,
      ...props
    } = this.props;
    const { loading } = this.state;

    const navigator = navigation.navigators[navigatorName];
    const scene = navigator.scenes[sceneName];

    const id = toId(navigatorName, sceneName);
    const pass = { loading, id };

    return (
      <>
        {overlay && (
          <Animated.View
            style={{
              position: 'absolute',
              height: '100%',
              width: '100%',
              overflow: 'hidden',
              transform: [
                {
                  translateY: scene.active.value.interpolate({
                    inputRange: [0, 0, 1],
                    outputRange: [height, 0, 0],
                  }),
                },
              ],
              backgroundColor: scene.active.value.interpolate({
                inputRange: [0, 1],
                outputRange: ['rgba(0,0,0,0)', 'rgba(0,0,0,0.4)'],
              }),
            }}
          />
        )}
        <Animated.View
          {...props}
          style={[
            {
              position: 'absolute',
              width: '100%',
              height: '100%',
              backgroundColor: 'white',
              overflow: 'hidden',
            },
            {
              transform: [
                {
                  translateY: scene.active.value.interpolate({
                    inputRange: [0, 1],
                    outputRange: [height, 0],
                  }),
                },
              ],
            },
            style,
          ]}
        >
          {typeof children === 'function' ? children(pass) : children}
        </Animated.View>
      </>
    );
  }
}
