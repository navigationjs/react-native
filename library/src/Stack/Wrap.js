import React, { Component } from 'react';
import { Animated, Dimensions, Platform } from 'react-native';
import { toId } from '../helpers';
import navigation from '../Navigation';

export default class Wrap extends Component {
  constructor(props) {
    super(props);

    const { navigator: navigatorName, scene: sceneName } = props;

    const navigator = navigation.navigators[navigatorName];
    const scene = navigator.scenes[sceneName];

    this.state = { loading: scene.active._value < 1 };
  }

  componentDidMount() {
    const { navigator: navigatorName, scene: sceneName } = this.props;

    const navigator = navigation.navigators[navigatorName];
    const scene = navigator.scenes[sceneName];

    scene.active.value.addListener(({ value }) => {
      if (value === 1) this.setState({ loading: false });
      else if (value === 0) this.setState({ loading: true });
    });
  }

  componentWillUnmount() {
    const { navigator: navigatorName, scene: sceneName } = this.props;

    const navigator = navigation.navigators[navigatorName];
    const scene = navigator.scenes[sceneName];

    scene.active.value.removeAllListeners();
  }

  render() {
    const {
      navigator: navigatorName,
      scene: sceneName,
      children,
      style,
      ...props
    } = this.props;
    const { loading } = this.state;

    const navigator = navigation.navigators[navigatorName];
    const scene = navigator.scenes[sceneName];

    const id = toId(navigatorName, sceneName);
    const pass = { loading, id };

    const { width } = Dimensions.get('window');

    return (
      <Animated.View
        {...props}
        style={[
          {
            position: 'absolute',
            width: '100%',
            height: '100%',
            backgroundColor: 'white',
            overflow: 'hidden',
          },
          {
            transform: [
              {
                translateX: scene.active.value.interpolate({
                  inputRange: [0, 1],
                  outputRange: [width, 0],
                }),
              },
              {
                translateX: scene.depth.value.interpolate({
                  inputRange: [0, 1],
                  outputRange: [0, -width / 5],
                }),
              },
            ],
          },
          style,
        ]}
      >
        {typeof children === 'function' ? children(pass) : children}
      </Animated.View>
    );
  }
}
