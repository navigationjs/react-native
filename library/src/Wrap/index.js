import React, { Component } from 'react';
import { View, BackHandler } from 'react-native';
import navigation, { Navigation } from '@navigationjs/core';

export default class Wrap extends Component {
  state = { lock: false };

  componentDidMount() {
    navigation.on(Navigation.EVENTS.LOCK, () => this.lock());
    navigation.on(Navigation.EVENTS.UNLOCK, () => this.unlock());

    this.backHandler = BackHandler.addEventListener(
      'hardwareBackPress',
      this.handleBackPress
    );
  }

  componentWillUnmount() {
    navigation.off(Navigation.EVENTS.LOCK);
    navigation.off(Navigation.EVENTS.UNLOCK);
    this.backHandler.remove();
  }

  handleBackPress = () => {
    const id = navigation.id();
    if (id) navigation.androidBack(id);
    return true;
  };

  lock = () => this.setState({ lock: true });
  unlock = () => this.setState({ lock: false });

  render() {
    const { children, ...props } = this.props;
    const { lock } = this.state;

    return (
      <View
        {...props}
        style={{ flex: 1 }}
        pointerEvents={lock ? 'none' : 'auto'}
      >
        {children}
      </View>
    );
  }
}
