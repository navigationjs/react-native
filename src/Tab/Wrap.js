import React, { Component } from 'react';
import { Animated, Dimensions } from 'react-native';
import { toId } from '../helpers';
import navigation from '../Navigation';

export default class Wrap extends Component {
  constructor(props) {
    super(props);

    const { navigator: navigatorName, scene: sceneName } = props;

    const navigator = navigation.navigators[navigatorName];
    const scene = navigator.scenes[sceneName];

    this.state = { loading: scene.active.loading };
  }

  componentDidMount() {
    const { navigator: navigatorName, scene: sceneName } = this.props;

    const navigator = navigation.navigators[navigatorName];
    const scene = navigator.scenes[sceneName];

    scene.active.value.addListener(() =>
      this.setState({ loading: scene.active.loading })
    );
  }

  render() {
    const {
      navigator: navigatorName,
      scene: sceneName,
      children,
      style,
      ...props
    } = this.props;
    const { loading } = this.state;

    const navigator = navigation.navigators[navigatorName];
    const scene = navigator.scenes[sceneName];

    const id = toId(navigatorName, sceneName);
    const pass = { loading, id };

    const { height } = Dimensions.get('window');

    return (
      <Animated.View
        {...props}
        style={[
          {
            position: 'absolute',
            width: '100%',
            height: '100%',
            backgroundColor: 'white',
            overflow: 'hidden',
          },
          {
            transform: [
              {
                translateY: scene.active.value.interpolate({
                  inputRange: [0, 1],
                  outputRange: [height, 0],
                }),
              },
            ],
          },
          style,
        ]}
      >
        {typeof children === 'function' ? children(pass) : children}
      </Animated.View>
    );
  }
}
