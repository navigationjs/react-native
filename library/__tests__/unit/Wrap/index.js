import React from 'react';
import { View } from 'react-native';
import events from '@railsmob/events';
import { render } from '@testing-library/react-native';
import navigation from '@navigationjs/core';
import Wrap from '../../../src/Wrap';

describe('Wrap', () => {
  it('should match snapshot', async () => {
    const { baseElement } = render(
      <Wrap>
        <View />
      </Wrap>
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should subscribe to lock and unlock', async () => {
    events.listeners = {};

    const { unmount } = render(
      <Wrap testID={'wrap'}>
        <View />
      </Wrap>
    );

    expect(events.listeners).toEqual({
      navigation_lock: [{ id: 'any', fn: expect.any(Function) }],
      navigation_unlock: [{ id: 'any', fn: expect.any(Function) }],
    });

    unmount();

    expect(events.listeners).toEqual({
      navigation_lock: [],
      navigation_unlock: [],
    });
  });

  it('should disable pointer events when navigation is locked', async () => {
    const { getByTestId } = render(
      <Wrap testID={'wrap'}>
        <View />
      </Wrap>
    );
    const component = getByTestId('wrap');
    expect(component.props.pointerEvents).toBe('auto');
    navigation.emit('lock');
    expect(component.props.pointerEvents).toBe('none');
    navigation.emit('unlock');
    expect(component.props.pointerEvents).toBe('auto');
  });
});
