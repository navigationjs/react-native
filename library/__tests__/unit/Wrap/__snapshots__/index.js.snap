// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrap should match snapshot 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View />
    </View>
  </View>
</View>
`;
